### PR TITLE
[ONNX] Inline prim::PythonOp for Autograd Function Export

### LINKED ISSUE #61813: ONNX export fails for trivial torch.autograd.Function
## ONNX export fails for trivial torch.autograd.Function

The torch.autograd.Function is very useful solution for training GPT3-like models with multi-GPU code.

It was used in very efficient implementation of transformer training developed by NVIDIA:
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/mpu/random.py#L319

The PyTorch ONNX trace can't correctly interpret simple usage of this function like this:
```
return CheckpointFunction.apply(function, *args)
```
It generates error for unknown Checkpoint.Function Python operator:
```
RuntimeError: ONNX export failed: Couldn't export Python operator CheckpointFunction
```

## To Reproduce

```python
import torch

class CheckpointFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, run_function, *args):
        return run_function(*args)

def checkpoint(function, *args):
    return CheckpointFunction.apply(function, *args)

class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
  
    def forward(self, input):
        def custom_forward(*inputs):
            x_ = inputs[0]
            return x_

        return checkpoint(custom_forward, input)

  
model = Net()
input = torch.zeros(2)
output = model(input, )
print(output)

with torch.no_grad():
    torch.onnx.export(model, (input, ), './check.onnx',)

```

The output from script:

```
Traceback (most recent call last):
  File "./apply_onnx.py", line 35, in <module>
    output = model(input, )
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "./apply_onnx.py", line 27, in forward
    return checkpoint(custom_forward, input)
  File "./apply_onnx.py", line 16, in checkpoint
    return CheckpointFunction.apply(function, *args)
AttributeError: type object 'CheckpointFunction' has no attribute 'apply'
root@nvdl-bsg-intel01:/mnt/nvdl/usr/piotrm/src/pytorch_multi_node# vim apply_onnx.py
root@nvdl-bsg-intel01:/mnt/nvdl/usr/piotrm/src/pytorch_multi_node# python3 ./apply_onnx.py
tensor([0., 0.])
Traceback (most recent call last):
  File "./apply_onnx.py", line 36, in <module>
    torch.onnx.export(model, (input, ), './check.onnx',)
  File "/opt/conda/lib/python3.8/site-packages/torch/onnx/__init__.py", line 271, in export
    return utils.export(model, args, f, export_params, verbose, training,
  File "/opt/conda/lib/python3.8/site-packages/torch/onnx/utils.py", line 88, in export
    _export(model, args, f, export_params, verbose, training, input_names, output_names,
  File "/opt/conda/lib/python3.8/site-packages/torch/onnx/utils.py", line 690, in _export
    proto, export_map = graph._export_onnx(
RuntimeError: ONNX export failed: Couldn't export Python operator CheckpointFunction

Defined at:
./apply_onnx.py(12): checkpoint
./apply_onnx.py(23): forward
/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py(1039): _slow_forward
/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py(1051): _call_impl
/opt/conda/lib/python3.8/site-packages/torch/jit/_trace.py(118): wrapper
/opt/conda/lib/python3.8/site-packages/torch/jit/_trace.py(127): forward
/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py(1051): _call_impl
/opt/conda/lib/python3.8/site-packages/torch/jit/_trace.py(1160): _get_trace_graph
/opt/conda/lib/python3.8/site-packages/torch/onnx/utils.py(373): _trace_and_get_graph_from_model
/opt/conda/lib/python3.8/site-packages/torch/onnx/utils.py(409): _create_jit_graph
/opt/conda/lib/python3.8/site-packages/torch/onnx/utils.py(445): _model_to_graph
/opt/conda/lib/python3.8/site-packages/torch/onnx/utils.py(676): _export
/opt/conda/lib/python3.8/site-packages/torch/onnx/utils.py(88): export
/opt/conda/lib/python3.8/site-packages/torch/onnx/__init__.py(271): export
./apply_onnx.py(36): <module>


Graph we tried to export:
graph(%input : Float(2, strides=[1], requires_grad=0, device=cpu)):
  %1 : Float(2, strides=[1], requires_grad=0, device=cpu) = ^CheckpointFunction(<function Net.forward.<locals>.custom_forward at 0x7f75e83ebdc0>)(%input) # ./apply_onnx.py:12:0
  return (%1)

```

## Expected behavior

Trivial model with one input/output should be generated by ONNX trace.

## Environment

```
PyTorch version: 1.9.0a0+c3d40fd
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.20.4
Libc version: glibc-2.31

Python version: 3.8.10 | packaged by conda-forge | (default, May 11 2021, 07:01:05)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-4.15.0-121-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: GPU 0: NVIDIA Graphics Device
Nvidia driver version: 470.42.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] nvidia-dlprof-pytorch-nvtx==1.2.0
[pip3] pytorch-quantization==2.1.0
[pip3] pytorch-transformers==1.1.0
[pip3] torch==1.9.0a0+c3d40fd
[pip3] torchtext==0.10.0a0
[pip3] torchvision==0.10.0a0
[conda] magma-cuda110             2.5.2                         5    local
[conda] mkl                       2019.5                      281    conda-forge
[conda] mkl-include               2019.5                      281    conda-forge
[conda] numpy                     1.20.3           py38h9894fe3_1    conda-forge
[conda] nvidia-dlprof-pytorch-nvtx 1.2.0                    pypi_0    pypi
[conda] pytorch-quantization      2.1.0                    pypi_0    pypi
[conda] pytorch-transformers      1.1.0                    pypi_0    pypi
[conda] torch                     1.9.0a0+c3d40fd          pypi_0    pypi
[conda] torchtext                 0.10.0a0                 pypi_0    pypi
[conda] torchvision               0.10.0a0                 pypi_0    pypi

```

## Additional context

cc @BowenBao @neginraoof

### PR DESCRIPTION
**Pull request by shubhambhokare1**
Add flag (inline_autograd) to enable inline export of model consisting of autograd functions. Currently, this flag should only be used in TrainingMode.EVAL and not for training. 

An example:

If a model containing ``autograd.Function`` is as follows
```
                class AutogradFunc(torch.autograd.Function):
                  @staticmethod
                  def forward(ctx, i):
                      result = i.exp()
                      result = result.log()
                      ctx.save_for_backward(result)
                      return result
```
Then the model is exported as
```
                graph(%0 : Float):
                  %1 : Float = ^AutogradFunc(%0)
                  return (%1)
```
If inline_autograd is set to True, this will be exported as
```
                graph(%0 : Float):
                  %1 : Float = onnx::Exp(%0)
                  %2 : Float = onnx::Log(%1)
                  return (%2)
```

If one of the ops within the autograd module is not supported, that particular node is exported as is mirroring ONNX_FALLTHROUGH mode

Fixes: #61813

### CODE REVIEW

**Review by BowenBao — CHANGES_REQUESTED** [review]
Thanks Shubham! This will make much better user experience when exporting autograd Function.

**Comment by BowenBao** [comment]
please rebase with master again to resolve conflict

File: test/onnx/test_autograd_funs.py
Comment (shubhambhokare1) [reply test/onnx/test_autograd_funs.py:25]: I had this test suite mimic test_utility_funs and the way model_to_graph is written in it. Any idea why it was setup that way?

File: test/onnx/test_autograd_funs.py
Comment (shubhambhokare1) [reply test/onnx/test_autograd_funs.py:118]: Although it looks similar, I was planning on adding more tests here based on future issues that autograd function export might face. I think the test cases in those might look pretty different

File: torch/csrc/jit/passes/onnx/autograd_function_process.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/autograd_function_process.cpp:29]: nit: newline

File: torch/csrc/jit/passes/onnx/autograd_function_process.cpp
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/autograd_function_process.cpp:18]: I wonder if it would be a possible use case to have nested autograd functions .. The it++ seems will skip the internal nodes. just curious

File: test/onnx/test_autograd_funs.py
Comment (BowenBao) [reply test/onnx/test_autograd_funs.py:41]: it could be outdated I guess, you can try calling the internal one directly.

File: torch/csrc/autograd/python_function.cpp
Comment (BowenBao) [inline torch/csrc/autograd/python_function.cpp:643]: Should this be controlled by `inline_autograd` flag?

File: torch/csrc/jit/passes/onnx/autograd_function_process.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/autograd_function_process.cpp:24]: Should this be controlled by `inline_autograd` flag?

File: torch/csrc/jit/passes/onnx/autograd_function_process.cpp
Comment (shubhambhokare1) [reply torch/csrc/jit/passes/onnx/autograd_function_process.cpp:18]: That's a great point. Adding a test case for this, do you think setting block iterator to the first node should work for this?

File: torch/onnx/utils.py
Comment (shubhambhokare1) [reply torch/onnx/utils.py:1201]: See: test_aten_unsupported

**Comment by BowenBao** [comment]
Please rebase with master to resolve conflict

File: test/onnx/test_autograd_funs.py
Comment (garymm) [inline test/onnx/test_autograd_funs.py:64]: Why is it helpful to test multiple opset versions?
If the logic being tested is independent of opset version, then I suggest we just test with a single opset version (maybe just use the default rather than specifying one).

File: torch/csrc/autograd/python_function.cpp
Comment (BowenBao) [inline torch/csrc/autograd/python_function.cpp:530]: Consider adding comment explaining some context. Essentially answering
* Where is original subgraph?
* Why do you need `subgraph_trace_outputs` and `unpack_output` (consider renaming)?
* What happens to original subgraph?

File: torch/onnx/symbolic_helper.py
Comment (shubhambhokare1) [reply torch/onnx/symbolic_helper.py]: removed previously, this is outdated i think

**Comment by BowenBao** [comment]
please rebase with master then check onnx CI, failures may be related.
```
       5 failed
         - test/onnx/test_utility_funs.py:1209 TestUtilityFuns_opset14.test_autograd_module_name
         - test/onnx/test_utility_funs.py:1178 TestUtilityFuns_opset14.test_autograd_onnx_fallthrough
         - test/onnx/test_utility_funs.py:1209 TestUtilityFuns_opset15.test_autograd_module_name
         - test/onnx/test_utility_funs.py:1178 TestUtilityFuns_opset15.test_autograd_onnx_fallthrough
         - test/onnx/test_utility_funs.py:489 TestUtilityFuns_opset15.test_constant_fold_div
```

File: torch/csrc/autograd/python_function.cpp
Comment (shubhambhokare1) [reply torch/csrc/autograd/python_function.cpp:530]: - unpack output is a variable to see whether if TupleUnpacking is needed

File: docs/source/onnx.rst
Comment (BowenBao) [inline docs/source/onnx.rst:508]: nit: indent different from others (this one is 2 spaces)

File: torch/csrc/jit/passes/onnx.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx.cpp:468]: complete todo?

File: torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp:49]: nit: this can be moved to outside and after this if block, so the other loop before if is not needed.

File: torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp:37]: move this out of `for (auto* n : subgraph->nodes())`, and iterate through `subgraph->outputs()` to register new outputs for `subblock`.

File: torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp:27]: nit: this messes with debugName potentially, but should be ok.

File: torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp:10]: nit: wonder if we can create subblock directly for autograd node? Might affect other existing jit passes maybe?

Comment (shubhambhokare1) [reply torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp:10]: will create an issue to optimize this conversion

File: torch/csrc/jit/passes/onnx.cpp
Comment (shubhambhokare1) [reply torch/csrc/jit/passes/onnx.cpp:468]: added more detailed warning

File: torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/pattern_conversion/autograd_function_process.cpp:10]: please comment with link to issue if possible...

**Review by BowenBao — APPROVED** [review]
LGTM. @albanD could you please take another look at the `python_function.cpp` part?

**Comment by shubhambhokare1** [comment]
Hey @eellison, could you provide a review, if possible, for this PR?

**Comment by BowenBao** [comment]
@malfet Can we move forward with this PR? It has been waiting for review for almost a month now. :)
IMO the change on JIT side is simple, and we could always revisit if it becomes a problem.

**Review by malfet — APPROVED** [review]

**Comment by BowenBao** [comment]
@malfet ping ..

**Comment by albanD** [comment]
Intern CI looks good!

**Comment by albanD** [comment]
@shubhambhokare1 could you do a rebase and fix the merge conflicts please? Once CI is green here, you can merge!

**Comment by shubhambhokare1** [comment]
@albanD Rebased, all CI is green except ```
Meta Internal-Only Changes Check```

**Comment by shubhambhokare1** [comment]
Any updates @albanD?

**Comment by BowenBao** [comment]
@malfet please help import and land... I think it's dragging too long and @shubhambhokare1 might need another rebase now after it was rebased last time for merge.

**Pull request closed without merging**
